### PR TITLE
Add physical memory allocator

### PR DESCRIPTION
**Pull request by cahirwpz**
Ready for merge with master.

### CODE REVIEW

**Comment by coodie** [comment]
Am i supposed to review the code, and accept merge pull request?

**Comment by cahirwpz** [comment]
You're supposed to review my changes. I'll merge PR myself.

File: vm_phys.c [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    vm_page_t *pg_it = TAILQ_FIRST(PG_FREEQ(seg, i));
+    while (pg_it) {
+      if (PG_START(pg_it) >= start && PG_END(pg_it) <= end) {
+        /* if segment is containted within (start, end) remove it from free
+         * queue */
+        pg_ptr = pg_it;
+        while (pg_ptr++ < pg_it + POW2(pg_it->order))
+          pg_ptr->flags |= VM_RESERVED;
+        TAILQ_REMOVE(PG_FREEQ(seg, pg_it->order), pg_it, freeq);
+        /* Go to start of list because it's corrupted */
```
Comment (cahirwpz) [inline vm_phys.c]: Could you explain why is it corrupted? This comment is worrisome.

Comment (coodie) [inline vm_phys.c]: You should not worry.

Usualy when you iterate over some structure, and modify it while traversing it, the iterator isn't reliable anymore. So the safest way to keep on traversing is to simply start doing it from the very start. And this is what i am doing here. So actually this comment is quite wrong, because list is totally ok, it's more like iterator is corrupted, or more like 'the way of iterating'.  In my opinion i should be able to remove iterator while iterating, because removing out of list doesn't change pointers of iterator, but pointers of elements adjacent to iterator, but i didn't want to rely on this, and (i don't know why) it didn't work.

In my opinion, the best idea would be to create set of elements to be deleted, but this means i'd need to use some kind of dynamic allocation to keep necessary information. 

So basically, i chose the safest solution.

File: vm_phys.c [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      } else if ((PG_START(pg_it) < start && PG_END(pg_it) > start) ||
+                 (PG_START(pg_it) < end && PG_END(pg_it) > end)) {
+        /* if segments intersects with (start, end) split it in half */
+        split_page(seg, pg_it);
+        /* Go to start of list because it's corrupted */
+        pg_it = TAILQ_FIRST(PG_FREEQ(seg, i));
+      } else {
+        pg_it = TAILQ_NEXT(pg_it, freeq);
+        /* if neither of two above cases is satisfied, leave in free queue */
+      }
```
Comment (coodie) [inline vm_phys.c]: Some grammar mistakes in comments in this function, also i'd change:
`/* Go to start of list because it's corrupted */`
to
`/* Go to start of list because iterator isn't reliable after removal */`

File: vm_phys.c [hunk 3]
```diff
@@ -0,0 +1,241 @@
+#include <vm_phys.h>
+#include <queue.h>
+#include <malloc.h>
+#include <libkern.h>
+
+#define POW2(x) (1 << (x))
+#define PG_START(pg) ((pg)->phys_addr)
+#define PG_END(pg) ((pg)->phys_addr + (1 << (pg->order)) * PAGESIZE)
+#define PG_FREEQ(seg, i) (&(seg)->free_queues[(i)])
```
Comment (coodie) [inline vm_phys.c]: Just a comment. I preffered older macro. I mean if you look at line:
`TAILQ_INSERT_HEAD(PG_FREEQ(seg, page->order), page, freeq);`
It's not immediately clear, what is happening here. And there is small opportunity of making a mistake, when choosing order of queue to insert page to. I'd rather prepare page, set up all necessary fields, and simply use:
`INSERT_FREE_PAGE(seg, page); `
Sure, we lose opportunity to insert at start or end, but algorithm isn't based on position of page in list, but on it's existence in list.

File: vm_phys.c (hunk 2 above)
Comment (cahirwpz) [inline vm_phys.c]: Ok. I'll include this change.

File: vm_phys.c (hunk 3 above)
Comment (cahirwpz) [inline vm_phys.c]: I'll reconsider it before merging with master.
